Sebastian/new sim mod version

sebwolf-de opened the pull request:
Hi, the new version of simmetrix (15.0-200330dev) introduces breaking changes to `void MS_setMeshSizePropagation`. It now has a new argument `int side`, which indicates into which direction the mesh size has to be propagated. By setting it to 2 it will be propagated into all directions.
I'll have to implement a switch at compile time to distinguish between the simmetrix versions, maybe with cmake ;-)

uphoffc committed:
Fix for SimModSuite 11/12

Commit:
Add initial support for cmake. Simmetrix support still missing.

Commit:
Add simmetrix support to cmake

sebwolf-de committed:
Update scons-tools

sebwolf-de committed:
Use new MeshSim API

Thomas-Ulrich commented:
Hi, maybe you can then merge the changes introduced here:
https://github.com/SeisSol/PUMGen/commit/7d5a8e1d3e3fe6bbec879c688f21c0c6d89d5a3b
(same problem of breaking changes that I fixed with my limited programming skills ;-))
Note that I do not use MeshSizePropagation anymore in my setup, but it was not getting me meshes as I would expect. (but maybe the feature has been improved now).

Thomas-Ulrich commented:
also about the change to scons-tools:
```
(['SimParasolid260', 'SimParasolid270', 'SimParasolid280', 'SimParasolid300', 'SimParasolid310'], None),
````
 It makes more sense to order them inversely (from newest to oldest).

sebwolf-de committed:
Merge remote-tracking branch 'origin/lukas/cmake' into sebastian/newSimModVersion

sebwolf-de committed:
Some adaptions for cmake

Thomas-Ulrich commented on src/input/SimModSuite.h:
```diff
@@ -440,18 +444,18 @@ void setCases(pGModel model, pACase &meshCase, pACase &analysisCase, MeshAttribu
-           int index = LUT[AnalysisAtt.faceBound[i]];
+        if (f.second!=0) {
+           logInfo(PMU_rank()) << "faceBound[" << f.first + 1 <<"] =" << f.second;
+           int index = LUT[f.second];
```
I think this lookup table could be removed now that you have the vector pairs.

sebwolf-de replied:
Thanks for your comment! I updated the whole part (also to get rid of the sprintf...)

Thomas-Ulrich replied:
ok looks good!

sebwolf-de committed:
cleanup

sebwolf-de committed:
Switch for Simmodeler 14

sebwolf-de committed:
Some edits to the cmake

sebwolf-de committed:
Cleanup

krenzland commented on src/input/AnalysisAttributes.cpp:
```diff
@@ -35,7 +34,7 @@ void AnalysisAttributes::set_BoundaryConditions() {
        split(tokens, line, ',');
        for(int i = 0; i < tokens.size(); i++) {
-           faceBound[std::atoi(tokens[i].c_str())-1] = 1;
+           faceBound.push_back(std::make_pair(std::atoi(tokens[i].c_str())-1, 1));
```
Note: https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-emplace.html would apply here
Same below

krenzland commented on src/input/AnalysisAttributes.h:
```diff
@@ -25,7 +27,7 @@
 class AnalysisAttributes {
   public:
-    int * faceBound;
+    std::vector<std::pair<int, int>> faceBound;
```
What is faceBound exactly? Maybe we could self-document this by using a vector of structs  instead of this approach?

krenzland commented on src/input/SimModSuite.h:
```diff
@@ -416,43 +418,39 @@ void setCases(pGModel model, pACase &meshCase, pACase &analysisCase, MeshAttribu
+       aBC[ufb] = AttCase_newModelAssoc(analysisCase,(pANode)iBC);
     }
 
     pGEntity face;
```
Here face is uninitialized - why isn't this done in the loop? In this case, it could be const as well

sebwolf-de replied:
Legacy :man_shrugging:

sebwolf-de replied:
But I've fixed it.

Thomas-Ulrich replied:
I thought that initializing over and over would be slower than initializing once.

krenzland reviewed (changes requested):
Just some small changes - otherwise it looks good. CMake stuff not reviewed because it's written mostly by me. Should be sound though.
At some point we should consider dropping support for scons for pumgen.

sebwolf-de committed:
Make faceBound more verbose

sebwolf-de commented:
Thanks for your review!

> At some point we should consider dropping support for scons for pumgen.

Might the the right time to start. Maybe with your recent cmake ci ;-)

sebwolf-de committed:
Merge branch 'xml' into sebastian/newSimModVersion

Thomas-Ulrich commented:
Could you document the cmake install process on the wiki page?
Do we need to specify all the 'APF_SIM_INCLUDE_DIR' and others variables?
```
di73yeq4@login02:/hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen/build_cmake> cmake .. -DSIMMETRIX_ROOT=/hppfs/work/pr63qo/di73yeq4/myLibs/SimModelerLib/14.0-190921
-- The C compiler identification is GNU 4.8.5
-- The CXX compiler identification is GNU 4.8.5
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Set build type to Release as none was supplied.
-- HDF5: Using hdf5 compiler wrapper to determine C configuration
-- Found HDF5: /dss/dsshome1/lrz/sys/spack/release/19.2/opt/x86_avx512/hdf5/1.8.20-intel-mhsivcc/lib/libhdf5.so;/usr/lib64/libpthread.so;/dss/dsshome1/lrz/sys/spack/release/19.2/opt/x86_avx512/libszip/2.1.1-gcc-v3th2mi/lib/libsz.so;/dss/dsshome1/lrz/sys/spack/release/19.2/opt/x86_avx512/zlib/1.2.11-gcc-ajxhsmr/lib/libz.so;/usr/lib64/libdl.so;/usr/lib64/libm.so (found version "1.8.20") found components:  C HL 
-- Found OpenMP_C: -fopenmp (found version "3.1") 
-- Found OpenMP_CXX: -fopenmp (found version "3.1") 
-- Found OpenMP: TRUE (found version "3.1")  
-- Found MPI_C: /dss/dsshome1/lrz/sys/intel/impi2019u6/compilers_and_libraries_2020.0.154/linux/mpi/intel64/lib/release/libmpi.so (found version "3.1") 
-- Found MPI_CXX: /dss/dsshome1/lrz/sys/intel/impi2019u6/compilers_and_libraries_2020.0.154/linux/mpi/intel64/lib/libmpicxx.so (found version "3.1") 
-- Found MPI: TRUE (found version "3.1")  
-- Could NOT find APF (missing: APF_INCLUDE_DIR APF_LIBRARIES) 
CMake Error at CMakeLists.txt:103 (set_target_properties):
  set_target_properties called with incorrect number of arguments.


CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
/hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen/APF_INCLUDE_DIR
   used as include directory in directory /hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen
/hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen/APF_SIM_INCLUDE_DIR
   used as include directory in directory /hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen
/hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen/APF_ZOLTAN_INCLUDE_DIR
   used as include directory in directory /hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen
/hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen/GMI_INCLUDE_DIR
   used as include directory in directory /hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen
/hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen/GMI_SIM_INCLUDE_DIR
   used as include directory in directory /hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen
/hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen/LION_INCLUDE_DIR
   used as include directory in directory /hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen
/hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen/MA_INCLUDE_DIR
   used as include directory in directory /hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen
/hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen/MDS_INCLUDE_DIR
   used as include directory in directory /hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen
/hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen/MTH_INCLUDE_DIR
   used as include directory in directory /hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen
/hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen/PCU_INCLUDE_DIR
   used as include directory in directory /hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen
APF_LIB
    linked by target "pumgen" in directory /hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen
APF_SIM_LIB
    linked by target "pumgen" in directory /hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen
APF_ZOLTAN_LIB
    linked by target "pumgen" in directory /hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen
GMI_LIB
    linked by target "pumgen" in directory /hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen
GMI_SIM_LIB
    linked by target "pumgen" in directory /hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen
LION_LIB
    linked by target "pumgen" in directory /hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen
MA_LIB
    linked by target "pumgen" in directory /hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen
MDS_LIB
    linked by target "pumgen" in directory /hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen
MTH_LIB
    linked by target "pumgen" in directory /hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen
PCU_LIB
    linked by target "pumgen" in directory /hppfs/work/pr63qo/di73yeq4/myLibs/PUMGen

-- Configuring incomplete, errors occurred!
```
Note that with scons it was rather straightforward:
```
hdf5Dir=$HDF5_BASE
pumiDIR='/hppfs/work/pr45fi/di73yeq4/myLibs/core/'
simetrixDir='/hppfs/work/pr63qo/di73yeq4/myLibs/SimModelerLib/14.0-190921'
scons simModSuite=yes mpiLib=mpich2 cc=mpicc cxx=mpicxx prefixPath=$hdf5Dir:$pumiDIR:$simetrixDir netcdf=no
```

sebwolf-de commented:
> Could you document the cmake install process on the wiki page?

I have this on my list

> Do we need to specify all the 'APF_SIM_INCLUDE_DIR' and others variables?

No, this should be found automatically by cmake. You can specify custom paths with 
-DCMAKE_PREFIX_PATH=~/...

Thomas-Ulrich commented on src/input/AnalysisAttributes.h:
```diff
@@ -15,18 +15,25 @@
 #include <string>
-#include <list>
-#include <map>
+#include <list>t
```
ups

sebwolf-de committed:
Fix Typo

Thomas-Ulrich commented:
Maybe you should try to compile your pull request :-p
```
mpicxx -o build/src/pumgen.o -c -Wall -ansi -std=c++11 -O2 -fopenmp -DLOG_LEVEL=2 -DNDEBUG -DUSE_SIMMOD -DBeforeSim14 -Isrc -I/dss/dsshome1/lrz/sys/spack/release/19.2/opt/x86_avx512/hdf5/1.8.20-intel-mhsivcc/include -I/hppfs/work/pr45fi/di73yeq4/myLibs/core/include -I/hppfs/work/pr63qo/di73yeq4/myLibs/SimModelerLib/14.0-190921/include -Isubmodules -I/dss/dsshome1/lrz/sys/spack/release/19.2/opt/x86_avx512/zlib/1.2.11-gcc-ajxhsmr/include -I/dss/dsshome1/lrz/sys/spack/release/19.2/opt/x86_avx512/libszip/2.1.1-gcc-v3th2mi/include src/pumgen.cpp
In file included from src/input/SimModSuite.h(46),
                 from src/pumgen.cpp(38):
src/input/AnalysisAttributes.h(18): warning #14: extra text after expected end of preprocessing directive
  #include <list>t
                 ^

In file included from src/pumgen.cpp(38):
src/input/SimModSuite.h(467): error: identifier "modelFaces" is undefined
      GFIter_delete(modelFaces);
                    ^

In file included from src/pumgen.cpp(38):
src/input/SimModSuite.h(558): error #140: too many arguments in function call
             MS_setMeshSizePropagation(meshCase,face,2,1,MeshAtt.MeshSizePropagationDistance,MeshAtt.MeshSizePropagationScalingFactor);
                                                                                             ^

compilation aborted for src/pumgen.cpp (code 2)
scons: *** [build/src/pumgen.o] Error 2
scons: building terminated because of errors.
```
Note that I get the last error with either True or False in BeforeSim14 (but may I wrongly use it)
```
scons simModSuite=yes mpiLib=mpich2 cc=mpicc cxx=mpicxx prefixPath=$hdf5Dir:$pumiDIR:$simetrixDir netcdf=no BeforeSim14=True
```

sebwolf-de committed:
Fix merge

Thomas-Ulrich commented:
It is possible that this code was not working at the end, and that I was appending the flag manually ;-)
```
if env['BeforeSim11']:
   env.Append(CPPDEFINES=['BeforeSim11'])
```

sebwolf-de committed:
Fix BeforeSim14 for scons

sebwolf-de commented:
Thanks for pointing this out. This came from the merge of the base branch, which I didn't thoroughly test.

Thomas-Ulrich commented:
Great! now it compiles. One last thing: Shouldn't be BeforeSim15 and not BeforeSim14?
14.0-190921 still uses the old fonction.
In other words, does BeforeSim14 should include 14, I think no.
But maybe this should be changed to simmodelerlib_lower_than_15.

sebwolf-de committed:
automatic switch for simmetrix version

sebwolf-de committed:
Add CI

sebwolf-de commented:
I changed the switch names and added a small CI (only without the simmetrix parts, as I am not sure, whether I can include their libraries into the ci due to licensing).

sebwolf-de committed:
Remove Debug Output

Thomas-Ulrich commented:
ok thx! No I think you can't include their libraries.

Thomas-Ulrich commented:
Can we merge?
